### PR TITLE
fix(@angular-devkit/schematics): VirtualDirEntry#subdirs & proper testing

### PR DESCRIPTION
**Pull request by ymeine**
- the getter `subdirs` in class `VirtualDirEntry` was wrongly implemented,
returning irrelevant data. Make some code common between `subdirs`
and `subfiles` for clarity and DRY respect.
- testing was not done properly either, since a `FileSystemTree` instance was used
(creating corresponding `FileSystemDirEntry` instances) in a test for `VirtualTree`
(and its corresponding model `VirtualDirEntry`). Despite the inheritance,
it wouldn't test it properly. Therefore copy this test as is to the proper file
`filesystem_spec.ts` and slightly adapt it in `virtual_spec.ts` to use the proper
implementations instead.

### CODE REVIEW

**Comment by ymeine** [comment]
@googlebot

File: packages/angular_devkit/schematics/src/tree/virtual.ts [hunk 1]
```diff
@@ -49,15 +49,20 @@ export class VirtualDirEntry implements DirEntry {
     return this._path == '/' ? null : this._tree.getDir(dirname(this._path));
   }
   get path() { return this._path; }
+
+  private _getDirectChildren(paths: Path[]) {
+    return paths
+      .filter(path => dirname(path) === this._path)
+      .map(path => basename(path));
```
Comment (hansl) [inline packages/angular_devkit/schematics/src/tree/virtual.ts]: This isn't proper logic. Your DRY has flaws in it.

If `files` is the following:

```js
[
  '/sub1/sub2/sub3/file1',
  '/sub1/sub4/sub5/file2',
  '/sub1/file3',
]
```

Listing `/sub1` should result in `[sub2, sub4, file3]`, but it won't using that new logic (it will list `[file1, file2, file3]`). There is a reason it wasn't DRY; the logic is different when listing files and dirs.

**Comment by hansl** [comment]
I think this PR is on the right track to fix the issue. It just needs to use prefixes instead of `dirname()`. Remember the `files` array is just an array of all file paths. So directories can jump from 1 sub to n sub.

File: packages/angular_devkit/schematics/src/tree/virtual.ts (hunk 1 above)
Comment (ymeine) [reply packages/angular_devkit/schematics/src/tree/virtual.ts]: Indeed, after diving into the code for the first and investigating hard, I didn't have sufficient resources anymore to design it properly I guess... I'm working on the fix right now :) (and of course on enhancing the current test cases accordingly)

**Comment by ymeine** [comment]
Fixing it properly was not that trivial finally, and led to more fixes and more test cases.

I added commits instead of squashing them for now, so you can review the changes only if you wish.

I added two different commits, since the last one refactors a bit the testing code I added but I'm not sure it suits your coding conventions. Not sure either for the comments I left in the code, even though I think they truly help understanding. Let me know what you think.

**Comment by hansl** [comment]
Hi @ymeine,

We reworked the schematics hosts to use the new Virtual FS. Could you rebase on top of it?

**Comment by ymeine** [comment]
@hansl Done, you might want to review the changes again even if it didn't change much. To summarize: 

- now when trying to visit a `FileSystemTree` instance given a path to a file an exception is thrown, checking that in the test
- the fix I did for `InMemoryFileSystemTreeHost` went away with the file, but I checked the new implementation in `SimpleMemoryHost#_isDir` and I added support for the edge case of the root path `/` (with according test)
